File: slowestKey.cpp
#include <iostream>
using namespace std;

#include <vector>

/*
	Intuition: Simple two-element sliding window traversal comparing 
			   the results according to the rules.

	RunTime: O(N)
	Space Complexity: O(1)
*/



class Solution {
public:
    char slowestKey(vector<int>& releaseTimes, string keysPressed) 
    {
        char result = keysPressed[0];
        int longest = releaseTimes[0];
        
        for (int i = 1; i < releaseTimes.size(); i++)
        {
            if ((releaseTimes[i] - releaseTimes[i - 1]) > longest)
            {
                longest = releaseTimes[i] - releaseTimes[i - 1];
                result = keysPressed[i];
            }
            
            else if ((releaseTimes[i] - releaseTimes[i - 1]) == longest)
            {
                result = (result > keysPressed[i]) ? result : keysPressed[i];
            }
        }
        
        return result;
    }
};

int main()
{
	vector<int> releaseTimes = {9,29,49,50};
	string keysPressed = "cbcd";

	Solution myObj;
	char answer = myObj.slowestKey(releaseTimes, keysPressed);

	cout << answer << endl;

	return 0;
}

File: slowestKey.py
from typing import List

"""
	Intuition: Simple sliding window of two elements
	RunTime: O(N), N = size of releaseTimes
	Space Complexity: O(1)
"""



class Solution:
    def slowestKey(self, releaseTimes: List[int], keysPressed: str) -> str:
        result = keysPressed[0]
        longest = releaseTimes[0]
        
        for i in range(1, len(releaseTimes)):
            if ((releaseTimes[i] - releaseTimes[i - 1]) > longest):
                longest = releaseTimes[i] - releaseTimes[i - 1]
                result = keysPressed[i]
        
            elif ((releaseTimes[i] - releaseTimes[i - 1]) == longest):
                if (result <= keysPressed[i]):
                    result = keysPressed[i]
        
        return result


releaseTimes = [9,29,49,50]
keysPressed = "cbcd"


myObj = Solution()
answer = myObj.slowestKey(releaseTimes,  keysPressed)
print(answer)
